Web Inspector: Show "Scroll" labels in the element tree for scrollable elements

patrickangle opened the pull request:
#### 5c6c14b2bd5558c1f301edd06c18de5e89e1b346
<pre>
Web Inspector: Show &quot;Scroll&quot; labels in the element tree for scrollable elements
<a href="https://webkit.org/b/243550">https://webkit.org/b/243550</a>
rdar://98519174

Reviewed by Devin Rousso.

Patch by Simon Fraser, with some rearranging by Patrick Angle.

Show &quot;Scroll&quot; labels on the document, and on scrollable elements in the Elements tab. We indicate document scrollability
on the `&lt;html&gt;` node, even in quirks mode where `document.scrollingElement()` is the `&lt;body&gt;`, unlike other browsers,
because otherwise you can&apos;t distinguish a scrollable document from a scrollable `&lt;body&gt;` on pages where the body is
independently scrollable.

&quot;Scrollable&quot; is a renderer flag just like &quot;Rendered&quot;, and DOMTreeElement.js makes a new badge for it just like &quot;flex&quot;
and &quot;grid&quot; badges.

Scrollable overflow depends on both style and layout, but doing invalidation at this level is tricky, so instead do
invalidation further downstream, when a renderer gains or loses a RenderLayer, and when a RenderLayerScrollableArea
updates its scrolling state after layout. For the FrameView, we invalidate when its scrollbars change.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Scrollable-expected.txt: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Scrollable.html: Added.
* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didAddOrRemoveScrollbarsImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didAddOrRemoveScrollbars):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol):
(WebCore::InspectorCSSAgent::didAddOrRemoveScrollbars):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::setContentsSize):
(WebCore::FrameView::addedOrRemovedScrollbar):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollInfoAfterLayout):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype.updateTitle):
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):

Canonical link: <a href="https://commits.webkit.org/254061@main">https://commits.webkit.org/254061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7129d2f3d4b6ce65b47f7afb04f5be8549cb1976

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87827 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31920 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97006 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/151501 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91795 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30282 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26342 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79917 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91750 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24452 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74549 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24428 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79410 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79557 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67275 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79611 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27950 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13375 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73347 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27933 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14390 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26073 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2852 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29625 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37302 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76180 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29514 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33666 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16892 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -77,6 +78,7 @@
 #include <wtf/Vector.h>
 #include <wtf/text/CString.h>
 #include <wtf/text/StringConcatenateNumbers.h>
+#include <wtf/text/TextStream.h>
```
NIT: Do we actually need this?

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -1006,6 +1021,8 @@ static RefPtr<JSON::ArrayOf<String /* Protocol::CSS::LayoutFlag */>> toProtocol(
     if (layoutFlags.contains(InspectorCSSAgent::LayoutFlag::Event))
         protocolLayoutFlags->addItem(Protocol::Helpers::getEnumConstantValue(Protocol::CSS::LayoutFlag::Event));
+    if (layoutFlags.contains(InspectorCSSAgent::LayoutFlag::Scrollable))
+        protocolLayoutFlags->addItem(Protocol::Helpers::getEnumConstantValue(Protocol::CSS::LayoutFlag::Scrollable));
```
We should put this right below `Rendered` to match the order in the protocol.

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -1068,6 +1085,11 @@ void InspectorCSSAgent::nodeHasLayoutFlagsChange(Node& node)
 
+void InspectorCSSAgent::didAddOrRemoveScrollbars(Node& node)
+{
+    didChangeRendererForDOMNode(node);
```
I think we should just call `nodeHasLayoutFlagsChange` directly.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -529,6 +529,14 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
         }
+
+        // COMPATIBILITY (macOS 13.0, iOS 16.0): `Scrollable` value for `CSS.LayoutFlag` did not exist yet.
+        if (InspectorBackend.Enum.CSS.LayoutFlag.Scrollable) {
```
This should probably be
```
InspectorBackend.Enum.CSS?.LayoutFlag?.Scrollable
```
just in case the target doesn't have a `CSS` domain or it's too old for `CSS.LayoutFlag` to exist.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -529,6 +529,14 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+
+        // COMPATIBILITY (macOS 13.0, iOS 16.0): `Scrollable` value for `CSS.LayoutFlag` did not exist yet.
+        if (InspectorBackend.Enum.CSS.LayoutFlag.Scrollable) {
+            contextMenu.appendCheckboxItem(WI.UIString("Scroll", "Title for a badge applied to DOM nodes that are a scrollable container."), () => {
```
We should move this above to match the order in the protocol.

At the very least this should probably be above Event 😅

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2206,8 +2214,9 @@ WI.DOMTreeElement.BadgeType = {
     Flex: "flex",
     Grid: "grid",
     Event: "event",
+    Scrollable: "scrollable",
```
ditto ([DOMTreeContentView.js](#diff-10809e8f4fe8e6b42f95bd1c9d63b82b278dae7bf4b52b5b3323e54b6def9dbe):535)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2038,6 +2038,10 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
             handleClick = this._handleEventBadgeClicked.bind(this);
             break;
+
+        case WI.DOMTreeElement.BadgeType.Scrollable:
```
What do we want to do for `handleClick`?  I think it'd be a bit odd for it to purely be presentational when all the other badges do something when clicked.  Perhaps we make it so that clicking toggles `overflow: hidden` (similar to how pressing `H` will toggle `visibility: hidden !important` via `__WebInspectorHideElement__`)?

also ditto ([DOMTreeContentView.js](#diff-10809e8f4fe8e6b42f95bd1c9d63b82b278dae7bf4b52b5b3323e54b6def9dbe):535)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2074,6 +2078,10 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
                 this._createBadge(WI.DOMTreeElement.BadgeType.Event);
                 break;
+
+            case WI.DOMNode.LayoutFlag.Scrollable:
```
ditto ([DOMTreeContentView.js](#diff-10809e8f4fe8e6b42f95bd1c9d63b82b278dae7bf4b52b5b3323e54b6def9dbe):535)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2206,8 +2214,9 @@ WI.DOMTreeElement.BadgeType = {
+    Scrollable: "scrollable",
 };
-WI.settings.enabledDOMTreeBadgeTypes = new WI.Setting("enabled-dom-tree-badge-types", [WI.DOMTreeElement.BadgeType.Flex, WI.DOMTreeElement.BadgeType.Grid, WI.DOMTreeElement.BadgeType.Event]);
+WI.settings.enabledDOMTreeBadgeTypes = new WI.Setting("enabled-dom-tree-badge-types", [WI.DOMTreeElement.BadgeType.Flex, WI.DOMTreeElement.BadgeType.Grid, WI.DOMTreeElement.BadgeType.Event, WI.DOMTreeElement.BadgeType.Scrollable]);
```
ditto ([DOMTreeContentView.js](#diff-10809e8f4fe8e6b42f95bd1c9d63b82b278dae7bf4b52b5b3323e54b6def9dbe):535)

dcrousso reviewed (approved):
awesome stuff! 🤩

i do have one lingering question on [DOMTreeElement.js](#diff-7f535a11b0098353d10e78b51224e1d5c17b62d1e61cbc0f0c69a50f9d453967):2042

smfr committed:
Web Inspector: Show "Scroll" labels in the element tree for scrollable elements
https://webkit.org/b/243550
rdar://98519174

Reviewed by Devin Rousso.

Patch by Simon Fraser, with some rearranging by Patrick Angle.

Show "Scroll" labels on the document, and on scrollable elements in the Elements tab. We indicate document scrollability
on the `<html>` node, even in quirks mode where `document.scrollingElement()` is the `<body>`, unlike other browsers,
because otherwise you can't distinguish a scrollable document from a scrollable `<body>` on pages where the body is
independently scrollable.

"Scrollable" is a renderer flag just like "Rendered", and DOMTreeElement.js makes a new badge for it just like "flex"
and "grid" badges.

Scrollable overflow depends on both style and layout, but doing invalidation at this level is tricky, so instead do
invalidation further downstream, when a renderer gains or loses a RenderLayer, and when a RenderLayerScrollableArea
updates its scrolling state after layout. For the FrameView, we invalidate when its scrollbars change.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Scrollable-expected.txt: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Scrollable.html: Added.
* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didAddOrRemoveScrollbarsImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didAddOrRemoveScrollbars):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol):
(WebCore::InspectorCSSAgent::didAddOrRemoveScrollbars):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::setContentsSize):
(WebCore::FrameView::addedOrRemovedScrollbar):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollInfoAfterLayout):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype.updateTitle):
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):

Canonical link: https://commits.webkit.org/254061@main
